File: src/sc_neurocore/accel/julia/neurons/sc_clipped_rational_recovery_map.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Julia retained clipped rational-recovery map

module SCClippedRationalRecoveryMapAccel

export simulate_trace

function simulate_trace(
    x0::Float64,
    y0::Float64,
    alpha::Float64,
    beta::Float64,
    j::Float64,
    x_threshold::Float64,
    clip_bound::Float64,
    n_steps::Int,
    current::Float64,
)
    values = (x0, y0, alpha, beta, j, x_threshold, clip_bound, current)
    all(isfinite, values) || throw(ArgumentError("SC rational-recovery inputs must be finite"))
    n_steps >= 0 || throw(ArgumentError("n_steps must be non-negative"))
    alpha > 0.0 && beta > 0.0 && clip_bound > 0.0 ||
        throw(ArgumentError("alpha, beta, and clip_bound must be positive"))
    abs(x0) <= clip_bound && abs(y0) <= clip_bound ||
        throw(ArgumentError("state exceeds clip_bound"))

    trace = Vector{Float64}(undef, n_steps)
    x = x0
    y = y0
    events = 0
    @inbounds for index in 1:n_steps
        x_previous = x
        field = x < 0.0 ? alpha * x : alpha * x / (1.0 + alpha * x)
        x_candidate = field + y + current + j
        y_candidate = y - beta * (x + 1.0)
        isfinite(x_candidate) && isfinite(y_candidate) ||
            throw(OverflowError("SC rational-recovery candidate became non-finite"))
        x = clamp(x_candidate, -clip_bound, clip_bound)
        y = clamp(y_candidate, -clip_bound, clip_bound)
        trace[index] = x
        if x >= x_threshold && x_previous < x_threshold
            events += 1
        end
    end
    return (trace = trace, events = events, xf = x, yf = y)
end

end

File: engine/src/neurons/sc_clipped_rational_recovery_map.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Retained clipped rational-recovery map

//! Count-neutral mirror of the retained project recurrence.

#[derive(Clone, Debug)]
pub struct SCClippedRationalRecoveryMapNeuron {
    pub x: f64,
    pub y: f64,
    pub alpha: f64,
    pub beta: f64,
    pub j: f64,
    pub x_threshold: f64,
    pub clip_bound: f64,
}

impl SCClippedRationalRecoveryMapNeuron {
    pub fn new() -> Self {
        Self {
            x: 0.0,
            y: 0.0,
            alpha: 3.0,
            beta: 0.001,
            j: 0.1,
            x_threshold: 1.0,
            clip_bound: 1_000_000.0,
        }
    }

    fn parameters_are_valid(&self) -> bool {
        [
            self.x,
            self.y,
            self.alpha,
            self.beta,
            self.j,
            self.x_threshold,
            self.clip_bound,
        ]
        .iter()
        .all(|value| value.is_finite())
            && self.alpha > 0.0
            && self.beta > 0.0
            && self.clip_bound > 0.0
            && self.x.abs() <= self.clip_bound
            && self.y.abs() <= self.clip_bound
    }

    /// Checked simultaneous update. Rejected input leaves state unchanged.
    pub fn try_step(&mut self, current: f64) -> Result<i32, &'static str> {
        if !self.parameters_are_valid() {
            return Err("invalid SC rational-recovery runtime state");
        }
        if !current.is_finite() {
            return Err("invalid SC rational-recovery current");
        }
        let x_previous = self.x;
        let field = if self.x < 0.0 {
            self.alpha * self.x
        } else {
            self.alpha * self.x / (1.0 + self.alpha * self.x)
        };
        let x_candidate = field + self.y + current + self.j;
        let y_candidate = self.y - self.beta * (self.x + 1.0);
        if !x_candidate.is_finite() || !y_candidate.is_finite() {
            return Err("SC rational-recovery candidate became non-finite");
        }
        let x_new = x_candidate.clamp(-self.clip_bound, self.clip_bound);
        let y_new = y_candidate.clamp(-self.clip_bound, self.clip_bound);
        let event = i32::from(x_new >= self.x_threshold && x_previous < self.x_threshold);
        self.x = x_new;
        self.y = y_new;
        Ok(event)
    }

    pub fn step(&mut self, current: f64) -> i32 {
        self.try_step(current).unwrap_or(0)
    }

    pub fn simulate(
        &mut self,
        n_steps: usize,
        current: f64,
    ) -> Result<(Vec<f64>, i64), &'static str> {
        let mut trace = Vec::with_capacity(n_steps);
        let mut events = 0_i64;
        for _ in 0..n_steps {
            events += i64::from(self.try_step(current)?);
            trace.push(self.x);
        }
        Ok((trace, events))
    }

    pub fn reset(&mut self) {
        self.x = 0.0;
        self.y = 0.0;
    }
}

impl Default for SCClippedRationalRecoveryMapNeuron {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn historical_orbit_is_retained() {
        let mut neuron = SCClippedRationalRecoveryMapNeuron::new();
        assert_eq!(neuron.try_step(0.0).unwrap(), 0);
        assert_eq!((neuron.x, neuron.y), (0.1, -0.001));
    }

    #[test]
    fn invalid_input_is_atomic() {
        let mut neuron = SCClippedRationalRecoveryMapNeuron::new();
        let before = (neuron.x, neuron.y);
        assert!(neuron.try_step(f64::NAN).is_err());
        assert_eq!((neuron.x, neuron.y), before);
    }
}
